Applications/Spire/Source/Ui/CalendarDatePicker.cpp

```diff
@@ -0,0 +1,441 @@
+    set(BackgroundColor(QColor::fromRgb(0x4B23A0))).
+    set(border(0, QColor::fromRgb(0, 0, 0, 0))).
+    set(TextColor(QColor::fromRgb(0xFFFFFF)));
+  set_style(*m_calendar_view, std::move(calendar_style));
+  m_model->connect_current_signal([=] (auto date) {
+    on_current(date);
+  });
+  update_calendar_model();
+  m_list_current_connection =
+    m_calendar_view->get_current_model()->connect_current_signal(
+      [=] (auto index) { on_list_current(index); });
+  m_calendar_view->connect_submit_signal([=] (const auto& model) {
+    on_submit(
+      std::any_cast<std::shared_ptr<LocalDateModel>>(model)->get_current());
+  });
+  layout->addWidget(m_calendar_view);
+}
+
+const std::shared_ptr<OptionalDateModel>&
+    CalendarDatePicker::get_model() const {
+  return m_model;
+}
+
+connection CalendarDatePicker::connect_submit_signal(
+    const SubmitSignal::slot_type& slot) const {
+  return m_submit_signal.connect(slot);
+}
+
+bool CalendarDatePicker::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == m_day_header && event->type() == QEvent::MouseButtonPress) {
+    m_calendar_view->setFocus();
+  }
+  if(event->type() == QEvent::KeyPress) {
+    auto e = static_cast<QKeyEvent*>(event);
+    if(watched == m_month_selector &&
+        (e->key() == Qt::Key_Up || e->key() == Qt::Key_Down)) {
+      qApp->sendEvent(m_calendar_view, e);
+    } else {
+      auto current_index = m_calendar_view->get_current_model()->get_current();
+      if(current_index) {
+        if(*current_index == 0 && e->key() == Qt::Key_Left) {
+          m_model->set_current(*m_model->get_current() - days(1));
+          return true;
+        } else if(*current_index == DISPLAYED_DAYS - 1 &&
+            e->key() == Qt::Key_Right) {
+          m_model->set_current(*m_model->get_current() + days(1));
+          return true;
+        }
+      }
+    }
+  }
+  return QWidget::eventFilter(watched, event);
+}
+
+int CalendarDatePicker::get_index(boost::gregorian::date day) const {
+  auto index = 0;
+  for(; index < m_calendar_view->get_list_model()->get_size(); ++index) {
+    if(m_calendar_view->get_list_model()->get<std::shared_ptr<
+        LocalDateModel>>(index)->get_current() == day) {
+      break;
+    }
+  }
+  return index;
+}
+
+void CalendarDatePicker::populate_calendar(const std::function<
+    void (int index, boost::gregorian::date day)> assign) {
+  auto displayed_month = m_month_selector->get_model()->get_current();
+  auto day = date(displayed_month.year(), displayed_month.month(), 1);
+  if(day.day_of_week() != 0) {
+    day += days(-day.day_of_week());
+  }
+  for(auto i = 0; i < DISPLAYED_DAYS; ++i) {
+    assign(i, day);
+    day += days(1);
+  }
+}
+
+void CalendarDatePicker::set_current_index(const boost::optional<int>& index) {
+  auto current_block =
+    shared_connection_block(m_list_current_connection);
+  m_calendar_view->get_current_model()->set_current(index);
+  m_calendar_view->get_selection_model()->set_current(index);
```

Comment:
I think it is needed anymore, because the selected can change with current in the updated listview.